🐛 [On call / 171] Source Salesforce: fixed the bug when `Bulk` fetch took all memory of kube pods

bazarnov opened the pull request:
## What
Resolving: https://github.com/airbytehq/oncall/issues/171

## How
- refactored `download_data` method to write the response by chunks into tmp file
- added `read_with_chunks` method to read the data from tmp file by chunks (lines)
- added `test_memory` unit_test to cover the case
- edited `Dockerfile` to support native installation of `pandas` and `numpy` frameworks

## 🚨 User Impact 🚨
No impact expected.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [X] Documentation updated
    - [X] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

bazarnov commented:
/test connector=connectors/source-salesforce

grubberr commented:
I have added `unit_tests/test_memory.py` file PLEASE REMOVE !!!

in really I still cannot see any difference

```
git checkout bazrnov/171-source-salesforce-oom-issue
pytest -s unit_tests/test_memory.py
first_size = 0.10 Mb, first_peak = 16.71 Mb

git checkout master
pytest -s unit_tests/test_memory.py
first_size = 0.10 Mb, first_peak = 16.71 Mb
```

bazarnov commented:
Can we adjust this test to use all memory in the system available? Let’s say 10gb file?
________________________________
From: Serhii Chvaliuk ***@***.***>
Sent: Tuesday, April 5, 2022 11:16:11 PM
To: airbytehq/airbyte ***@***.***>
Cc: Baz ***@***.***>; Assign ***@***.***>
Subject: Re: [airbytehq/airbyte] 🐛 [On call / 171] Source Salesforce: fixed the bug when `Bulk` fetch took all memory of kube pods (PR #11692)


I have added unit_tests/test_memory.py file PLEASE REMOVE !!!

in really I still cannot see any difference

git checkout bazrnov/171-source-salesforce-oom-issue
pytest -s unit_tests/test_memory.py
first_size = 0.10 Mb, first_peak = 16.71 Mb

git checkout master
pytest -s unit_tests/test_memory.py
first_size = 0.10 Mb, first_peak = 16.71 Mb


—
Reply to this email directly, view it on GitHub<https://github.com/airbytehq/airbyte/pull/11692#issuecomment-1089286222>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFPMHGRF7R3I2XXWH7NG24DVDSNQXANCNFSM5SPH4HPA>.
You are receiving this because you were assigned.Message ID: ***@***.***>

bazarnov commented:
/test connector=connectors/source-salesforce

grubberr commented:
BTW is it good idea to store data in intermediate temporary file ?
I can propose as an option more simpler solution without temp file and pandas library:
https://gist.github.com/grubberr/640bf8d1d7d9b83418cc0b8c21b70cbb

bazarnov commented:
> BTW is it good idea to store data in intermediate temporary file ? I can propose as an option more simpler solution without temp file and pandas library: https://gist.github.com/grubberr/640bf8d1d7d9b83418cc0b8c21b70cbb

I don't mind of using this approach, however there is no guarantee that the issue will not appear again, since the response could be a huge size. What are your concerns of using tmp files here?

girarda commented on airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:
```diff
@@ -252,18 +255,56 @@ def filter_null_bytes(self, s: str):
-        res = s.replace("\x00", "")
+        replace_list = ['"', "\x00", "\0", "\00"]
+        for rep in replace_list:
+            res = s.replace(rep, "")
```
this will overwrite the replace from the previous elements.
It might be worth adding a test case for a string with all 4 null bytes and make sure they are all replaced

grubberr replied:
sorry I did not looked closely first time to this place
```
>>> "\x00" == "\0" == "\00"
True
```

?

girarda commented on airbyte-integrations/connectors/source-salesforce/integration_tests/bulk_error_test.py:
```diff
@@ -87,6 +87,4 @@ def test_failed_jobs_with_successful_switching(caplog, input_sandbox_config, str
         with caplog.at_level(logging.WARNING):
             loaded_record_ids = set(record["Id"] for record in stream.read_records(sync_mode=SyncMode.full_refresh))
-        for i, log_message in enumerate(log_messages, 1):
-            assert log_message in caplog.records[-i].message
```
why is this assert not needed anymore?

bazarnov replied:
Returned back. Was lost during experiments.

girarda commented on airbyte-integrations/connectors/source-salesforce/unit_tests/test_memory.py:
```diff
@@ -0,0 +1,46 @@
+        (1000, 0.4, 1),
+        (10000, 1, 2),
+        (100000, 4, 7),
+        # (200000, 6, 12), # takes a long time
```
just to confirm - it takes a long time but it passes?

bazarnov replied:
yes)

girarda commented on airbyte-integrations/connectors/source-salesforce/unit_tests/test_memory.py:
```diff
@@ -0,0 +1,46 @@
+        fs, fp = tracemalloc.get_traced_memory()
+        first_size_in_mb, first_peak_in_mb = fs / 1024**2, fp / 1024**2
+
+        assert True if first_size_in_mb < first_size else False
```
nit: this could be simplified to
`assert first_size_in_mb < first_size`

bazarnov replied:
Good point, thanks.

girarda commented on airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:
```diff
@@ -252,18 +255,56 @@ def filter_null_bytes(self, s: str):
+        if os.path.isfile(tmp_file):
+            return tmp_file
+        else:
+            self.logger.error(f"Could not verify binary data for stream `{self.name}`!")
```
should this raise an exception?

bazarnov replied:
Corrected.

girarda commented on airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py:
```diff
@@ -252,18 +255,56 @@ def filter_null_bytes(self, s: str):
+                yield from []
+
+        # clean temporarily binary file, after data is read
+        if os.path.isfile(path):
```
we should wrap this block in a finally to ensure it gets executed even if an exception is thrown

bazarnov replied:
Done.

bazarnov commented:
/test connector=connectors/source-salesforce

girarda reviewed (approved):
small question around the TmpFileIOError's constructor, but looks good otherwise

bazarnov commented on airbyte-integrations/connectors/source-salesforce/source_salesforce/exceptions.py:
```diff
@@ -13,3 +23,8 @@ class TypeSalesforceException(SalesforceException):
+
+class TmpFileIOError(Error):
+    def __init__(self, msg: str, err: str = None):
+        super().__init__(self.logger.fatal(f"{msg} for job: {msg}. Error: {err}"))
```
Checked. Calling `super()` was not necessary in this case, so removed it.

bazarnov commented:
/publish connector=connectors/source-salesforce

bazarnov commented:
/publish connector=connectors/source-salesforce
